projects/sdk/src/lib/silo/Withdraw.ts

```diff
@@ -0,0 +1,128 @@
+import { DepositCrate, Silo } from "../silo";
+import { sortCratesBySeason } from "../silo.utils";
+
+export class Withdraw {
+  static sdk: BeanstalkSDK;
+
+  constructor(sdk: BeanstalkSDK) {
+    Withdraw.sdk = sdk;
+  }
+
+  /**
+   * Initates a withdraw from the silo. The `token` specified dictates which silo to withdraw
+   * from, and therefore is limited to only whitelisted assets.
+   * Behind the scenes, the `amount` to be withdrawn must be taken from individual
+   * deposits, aka crates. A user's deposits are not summarized into one large bucket, from
+   * which we can withdraw at will. Each deposit is independently tracked, so each withdraw must
+   * calculate how many crates it must span to attain the desired `amount`.
+   * @param token The whitelisted token to withdraw. ex, BEAN vs BEAN_3CRV_LP
+   * @param amount The desired amount to withdraw. Must be 0 < amount <= total deposits for token
+   * @returns Promise of Transaction
+   */
+  async withdraw(token: Token, amount: TokenValue): Promise<ContractTransaction> {
+    Withdraw.sdk.debug("silo.withdraw()", { token, amount });
+    if (!Withdraw.sdk.tokens.siloWhitelist.has(token)) {
+      throw new Error(`Withdraw error; token ${token.symbol} is not a whitelisted asset`);
+    }
+
+    const { deposited } = await Withdraw.sdk.silo.getBalance(token);
+    Withdraw.sdk.debug("silo.withdraw(): deposited balance", { deposited });
+
+    if (deposited.amount.lt(amount)) {
+      throw new Error("Insufficient balance");
+    }
+
+    const season = await Withdraw.sdk.sun.getSeason();
+
+    const withdrawData = this.calculateWithdraw(token, amount, deposited.crates, season);
+    Withdraw.sdk.debug("silo.withdraw(): withdrawData", { withdrawData });
+
+    const seasons = withdrawData.crates.map((crate) => crate.season.toString());
+    const amounts = withdrawData.crates.map((crate) => crate.amount.toBlockchain());
+
+    let contractCall;
+
+    if (seasons.length === 0) {
+      throw new Error("Malformatted crates");
+    }
+
+    if (seasons.length === 1) {
+      Withdraw.sdk.debug("silo.withdraw(): withdrawDeposit()", { address: token.address, season: seasons[0], amount: amounts[0] });
+      contractCall = Withdraw.sdk.contracts.beanstalk.withdrawDeposit(token.address, seasons[0], amounts[0]);
+    } else {
+      Withdraw.sdk.debug("silo.withdraw(): withdrawDeposits()", { address: token.address, seasons: seasons, amounts: amounts });
+      contractCall = Withdraw.sdk.contracts.beanstalk.withdrawDeposits(token.address, seasons, amounts);
+    }
+
+    return contractCall;
+  }
+
+  /**
+   * This methods figures out which deposits, or crates, the withdraw must take from
+   * in order to reach the desired amount. It returns extra information that may be useful
+   * in a UI to show the user how much stalk and seed they will forfeit as a result of the withdraw
+   */
+  calculateWithdraw(token: Token, amount: TokenValue, crates: DepositCrate[], season: number) {
```

Comment:
What's the use case for this (and the pickCrates) being publicly accessible? Do we surface that in the UI somewhere? If so, might make sense to add to `examples`.